README.md
#Funimation XBMC Plugin

An XBMC plugin for streaming videos from funimation.com.

If you're interested in helping out, just drop me an email or send a pull
request. Patches and (constructive) input are always welcome.

PS3 API Information
-------------------
This just has some info about the API the PS3 app uses. The API backend appears to be using the [CodeIgniter](http://www.codeigniter.com/) framework.

API endpoints
- Shows `/feeds/ps/shows`
- Episodes `/feeds/ps/videos`
- Featured Shows `/feeds/ps/featured`
- Login `/feeds/ps/login.json?v=2`
- Logout `/feeds/ps/logout`
- Search `/feeds/ps/search`
- Main Screen Assests `/feeds/ps/getDeviceAssets`

---
#####Request info
`User-agent: Sony-PS3`

All requests have to have the query param ut=`user_type` from the login results, like ut=FunimationSubscriptionUser.

---
Here are the query params and values that I have found so far.
####shows
| Param        | Values                                                   |
| ------------ | ----------------------------                             |
| ut           | FunimationSubscriptionUser &#124; FunimationUser         |
| limit        | `int`                                                    |
| offset       | `int`                                                    |
| sort         | SortOptionLatestSubscription &#124; SortOptionLatestFree |
| first-letter | [a-z] &#124; non-alpha                                   |
| filter       | FilterOptionSimulcast                                    |

####Episodes
| Param        | Values                                           |
| ------------ | ----------------------------                     |
| ut           | FunimationSubscriptionUser &#124; FunimationUser |
| limit        | `int`                                            |
| offset       | `int`                                            |
| show_id      | `int`                                            |

####Featured
| Param        | Values                                           |
| ------------ | ----------------------------                     |
| ut           | FunimationSubscriptionUser &#124; FunimationUser |

####Search
| Param        | Values                                           |
| ------------ | ----------------------------                     |
| ut           | FunimationSubscriptionUser &#124; FunimationUser |
| search       | `str`                                            |


---
####Login
post body 
playstation_id is always blank for some reason. The `Content-Type` header must be set to `application/json` or you get a `Disallowed Key Characters` error.
```
{
    "username": "User",
    "password": "Password",
    "playstation_id": ""
}
```

returns
user_age appears to be off by one year...
```
{
    "user_id": "1234567",
    "user_type": "FUNIMATION_SUBSCRIPTION_USER",
    "user_birthday": "01-23-1991",
    "user_age": 22,
    "country": "US"
}
```

---
####Shows
```
[{
    "asset_id": "7556822",
    "pubDate": "08\/27\/2013",
    "series_name": "Cowboy Bebop",
    "link": "http:\/\/www.funimation.com\/cowboy-bebop",
    "series_description": "Explore the galaxy in this undeniably hip series that inspired a generation \u2013 and redefined anime as an indisputable art form. The Bebop crew is just trying to make a buck, and they\u2019re the most entertaining gang of bounty hunters in the year 2071.",
    "season_count": "1",
    "episode_count": 26,
    "genres": "Action,Sci Fi,Space",
    "simulcast": "0",
    "popularity": "4842",
    "official_marketing_website": "http:\/\/www.funimation.com\/cowboy-bebop",
    "latest_video_free": {
        "video_id": "57970",
        "release_date": "1417759200",
        "title": "Episode 4 (Dub)"
    },
    "latest_video_free_release_date": "1417759200",
    "latest_video_subscription": {
        "video_id": "58047",
        "release_date": "1417500000",
        "title": "Episode 26 (Dub)"
    },
    "latest_video_subscription_release_date": "1417500000",
    "show_rating": "TV-14",
    "active_hd_1080": "1",
    "thumbnail_small": "http:\/\/www.funimation.com\/admin\/uploads\/default\/shows\/show_thumbnail\/1_thumbnail\/CBY_thumb.jpg",
    "thumbnail_medium": "http:\/\/www.funimation.com\/admin\/uploads\/default\/shows\/show_mobile\/similar_shows\/CBY_mobile.jpg",
    "thumbnail_large": "http:\/\/www.funimation.com\/admin\/uploads\/default\/shows\/show_mobile\/1_device_show\/CBY_mobile.jpg",
    "poster_art": "http:\/\/www.funimation.com\/admin\/uploads\/default\/shows\/show_tv\/1_device_show\/CBY_tv_new.jpg",
    "contactLink": "http:\/\/www.funimation.com\/support",
    "display_order": 0,
    "element_position": 16,
    "languages": "Dub, Sub",
    "quality": "HD (1080)"
},]
```

---
####Episodes
```
{
  "videos" : [
    {
      "sequence" : "1",
      "quality" : "HD (1080)",
      "genre" : "Action,Sci Fi,Space",
      "tv_or_move" : "tv",
      "pubDate" : "11\/17\/2014",
      "url" : "asteroid-blues",
      "duration" : 1496,
      "description" : "Spike and Jet head to Tijuana to track down an outlaw smuggling a dangerous drug known as blood-eye.  Jet wants the bounty, but Spike has eyes for a far prettier prize.",
      "thumbnail_large" : "http:\/\/www.funimation.com\/admin\/uploads\/default\/recap_thumbnails\/7556822\/home_spotlight\/CBY0001.jpg",
      "closed_caption_location" : "",
      "show_name" : "Cowboy Bebop",
      "number" : 1,
      "releaseDate" : "2014\/11\/18",
      "featured" : "false",
      "highdef" : "true",
      "simulcast" : "0",
      "closed_captioning" : "0",
      "element_position" : 0,
      "display_order" : 0,
      "thumbnail" : "CBY0001.jpg",
      "has_subtitles" : "false",
      "extended_title" : "Cowboy Bebop : Episode 1 - Asteroid Blues (Dub)",
      "thumbnail_small" : "http:\/\/www.funimation.com\/admin\/uploads\/default\/recap_thumbnails\/7556822\/playlist\/CBY0001.jpg",
      "thumbnail_url" : "http:\/\/www.funimation.com\/admin\/uploads\/default\/recap_thumbnails\/7556822\/videos_spotlight\/CBY0001.jpg",
      "aips" : [
        "152.194",
        "758.091",
        "1009.425"
      ],
      "video_type" : "Episode",
      "dub_sub" : "dub",
      "popularity" : "4630",
      "video_url" : "http:\/\/wpc.8c48.edgecastcdn.net\/038C48\/SV\/480\/CBYENG0001\/CBYENG0001-480-,750,1500,2000,2500,4000,K.mp4.m3u8?9b303b6c62204a9dcb5ce5fdc007bb4104d2360e7fab15f96cf6e5f448e60394b08e9a09c51f786234bcd67b13e585ce7177b0a798e431722cf8aff5e61f5d2b274522d95cd9478b5d33afecce35f2cf1250c37930429ce355f94462",
      "asset_id" : "57967",
      "show_id" : "7556822",
      "title" : "Asteroid Blues",
      "language" : "en",
      "funimation_id" : "CBYENG0001",
      "rating" : "TV-14",
      "rating_system" : "tvpg",
      "thumbnail_medium" : "http:\/\/www.funimation.com\/admin\/uploads\/default\/recap_thumbnails\/7556822\/1_thumbnail\/CBY0001.jpg"
    }
  ]
}
```


changelog.txt
v1.1.3
- Fixed clear cookie fail message

v1.1.2
- Fixed module import error

v1.1.1
- Removed unused dependeny
- Fixed addon ID

v1.1.0
- Almost a complete rewrite
- Using the PS3 app API
- More menus and better quality images
- Now displays, episode and show details
- Removed caching

v1.0.3
- Added a check to see if user is a subscriber
- The clear cache option now properly clears the StorageServer cache

v1.0.2
- Removed some unused strings
- Removed 'optional="true"' from addon.xml because it does nothing

v1.0.1
- Added check for addon userdata directory

v1.0.0
- Initial release

v0.2.1
- Changed error list item to be more generic

v0.2.0
- Fixed getting movies
- Added setting streaming info
- Cleaned up some code

v0.1.2
- Added caching
- Made common plugin cache and optional import
- Added option to disable caching

v0.1.1
- Rewrote almost everything
- Removed model/controller classes

v0.1.0
- Rewrote api classes
- Removed user model/controller
- Cleaned up code

v0.0.9
- Added setting for adding shows to a black list
- Added http_client param to user_controller

v0.0.8
- Added call to login when getting the list of shows

v0.0.7
- Moved UserController initialization to speed up addon start

v0.0.6
- Added asynchronous login when the addon is run

v0.0.5
- Added movies, trailers, and clips

v0.0.4
- Implemented search
- Implemented caching
- Added setting to select which genres to show

v0.0.3
- Rewrote the plugin to not use xbmcswift2

v0.0.2
- Created model/controllers
- Added controls for caching.

v0.0.1
- Initial version


resources/lib/models.py

class TemplateModel(object):
    _common_fields = [
        'asset_id',
        'display_order',
        'element_position',
        'popularity',
        'pubDate',
        'quality',
        'simulcast',
        'thumbnail_large',
        'thumbnail_medium',
        'thumbnail_small',
    ]

    _fields = []

    def __init__(self, **kwargs):
        self._fields.extend(self._common_fields)
        for k, v in kwargs.iteritems():
            if k in self._fields:
                try:
                    setattr(self, k, v)
                except Exception, e:
                    print(k, v)
                    raise e

    @property
    def icon(self):
        return self.thumbnail_large

    @property
    def label(self):
        return NotImplementedError()

    @property
    def query_string(self):
        raise NotImplementedError()

    @property
    def stream_info(self):
        if self.video_quality == 4000:
            w, h = 1920, 1080
        elif self.video_quality == 3500:
            w, h = 1080, 720
        else:
            w, h = 640, 480
        aspect = w / float(h)

        si = {
            'codec': 'mp4',
            'aspect': aspect,
            'width': w,
            'height': h,
            'duration': self.duration
        }

        return si

    def get(self, key, default=None):
        return getattr(self, key, default)

    # not sure if i like this, might show to much info
    def __repr__(self):
        return repr(self.__dict__)


class Show(TemplateModel):
    _fields = [
        'series_name',
        'link',
        'series_description',
        'season_count',
        'episode_count',
        'genres',
        'official_marketing_website',
        'latest_video_free',
        'latest_video_free_release_date',
        'latest_video_subscription',
        'latest_video_subscription_release_date',
        'show_rating',
        'active_hd_1080',
        'poster_art',
        'contactLink',
        'languages',
    ]

    @property
    def thumbnail(self):
        return self.poster_art

    @property
    def label(self):
        return self.series_name

    @property
    def label2(self):
        return '[]'

    @property
    def vtypes(self):
        return 'episodes'

    @property
    def info(self):
        video_info = {
            'genre': self.genres,
            'votes': self.popularity,
            'plot': self.series_description,
            'episode': self.episode_count,
            'aired': self.pubDate.strftime('%Y-%m-%d'),
            'year': self.pubDate.strftime('%Y'),
        }

        return video_info

    @property
    def query_string(self):
        qry = {
            'show_id': self.asset_id,
            'folder': 'true',
            'path': '/show/episodes',
            'get': 'episodes',
            'vtypes': self.vtypes,
        }

        return qry

    @property
    def stream_info(self):
        raise NotImplementedError()

    @property
    def sub(self):
        raise NotImplementedError()

    @property
    def dub(self):
        raise NotImplementedError()


class Episode(TemplateModel):
    _fields = [
        'aips',
        'closed_caption_location',
        'closed_captioning',
        'description',
        'dub_sub',
        'duration',
        'extended_title',
        'featured',
        'funimation_id',
        'genre',
        'has_subtitles',
        'highdef',
        'language',
        'number',
        'rating',
        'rating_system',
        'releaseDate',
        'sequence',
        'show_id',
        'show_name',
        'simulcast',
        'thumbnail_url',
        'title',
        'tv_or_move',
        'url',
        'video_type',
        'video_url',
    ]

    @property
    def label(self):
        if self.number:
            lbl = '%d. %s (%s)' % (self.number, self.title, self.dub_sub)
        else:
            lbl = '%s (%s)' % (self.title, self.dub_sub)
        return lbl

    @property
    def thumbnail(self):
        return self.thumbnail_url

    @property
    def sub(self):
        return self.dub_sub == 'sub'

    @property
    def dub(self):
        return self.dub_sub == 'dub'

    @property
    def info(self):
        video_info = {
            'duration': self.duration / 60,
            'episode': self.episode_number,
            'votes': self.votes,
            'genre': self.genre,
            'plot': self.description,
            'aired': self.pubDate.strftime('%Y-%m-%d'),
            'tvshowtitle': self.show_title,
        }

        return video_info

    @property
    def video_quality(self):
        if self.quality == 'HD (1080)':
            return 4000
        elif self.quality == 'HD (720)':
            return 3500
        else:
            return 2000

    @property
    def query_string(self):
        qry = {
            'id': self.asset_id,
            'path': '/show/episodes',
            'videoid': self.funimation_id,
        }

        return qry
